Component library support

chrisdp opened the pull request:
This is very early support for component libraries. I am interested in feedback as this is the first time I have worked on static file hosting and file manipulation.

chrisdp commented:
[CompLibTestChannel.zip](https://github.com/TwitchBronBron/vscode-brightscript-language/files/3244977/CompLibTestChannel.zip)
This is a sample channel you can use to test this. It is currently set up so that you open VS Code with the compWorkspace.code-workspace file to open it as a work space of two nested work spaces but you also could just have the launch config one level higher in the folder structure. 

I have it done this way to simulate my work environment where we have a demo channel and the comp lib checked out in parallel with there own .vscode fold with launch configs and such.

NathanGillis commented:
I tested this PR out and it works!!!! Nice Job @chrisdp. Breakpoints in component library NICE.

TwitchBronBron commented on package.json:
```diff
@@ -156,6 +157,30 @@
                                     "${workspaceFolder}"
                                 ]
                             },
+                            "componentLibraryPort": {
```
Let's make this plural like the other ones. `componentLibrariesPort`.

TwitchBronBron commented on src/BrightScriptDebugSession.ts:
```diff
@@ -287,6 +306,157 @@ export class BrightScriptDebugSession extends DebugSession {
+                this.convertBreakpointPaths(componentLibrary.rootDir, componentLibrary.rootDir);
+                await this.addBreakpointStatements(stagingFolder, componentLibrary.rootDir);
+
+                for (let filePath of paths) {
```
I know it makes this code a little more complicated, but we should consider running each iteration in parallel (not actual parallel, but the node style). So something like this: 

```typescript
await Promise.all(paths.map(async (filePath)=>{
    ....the for loop body
{));
```

This will allow things to complete a bit faster because we are doing work as soon as we are free, instead of iterating through each path in order.

TwitchBronBron commented on src/BrightScriptDebugSession.ts:
```diff
@@ -287,6 +306,157 @@ export class BrightScriptDebugSession extends DebugSession {
+                            });
+
+                            // Rename the brs files to include the postfix name spacing tag
+                            fs.rename(filePath, path.join(stagingFolder, relativePath), function(err) {
```
These callbacks are not awaited, so technically we could encounter race conditions where the files haven't been totally renamed yet. 

In addition to the previous request to make this for loop async/parallel, you should consider using a promise-based rename function, and awaiting it. Check out the `move` function from the [fs-extra](https://github.com/jprichardson/node-fs-extra/blob/master/docs/move.md) package. You can do something like this:

```typescript
await fsExtra.move(source, dest);
```

TwitchBronBron commented on src/BrightScriptDebugSession.ts:
```diff
@@ -287,6 +306,157 @@ export class BrightScriptDebugSession extends DebugSession {
+        if (componentLibraries && componentLibrariesOutDir) {
+            this.componentLibrariesOutDir = componentLibrariesOutDir;
+            this.componentLibrariesStagingDirPaths = [];
+            let replace = require('replace-in-file');
```
Move this to the top of the file with the other imports, and use the typescript import syntax. 
```typescript
import * as replaceInFile from 'replace-in-file';
```

chrisdp replied:
Moving to this type of import seems to be giving me a problem. 
![image](https://user-images.githubusercontent.com/9591618/59078645-6e738280-88b6-11e9-9894-2ee93534629c.png)

chrisdp replied:
Not sure really why that is though.

TwitchBronBron commented on src/BrightScriptDebugSession.ts:
```diff
@@ -287,6 +306,157 @@ export class BrightScriptDebugSession extends DebugSession {
+                            relativePath = path.join(parsedPath.dir, newFileName);
+
+                            // Update all the file name references in the library to the new file names
+                            replace.sync({
```
Use an async promise-returning version of this function

```typescript
await replace(...yourStuff);
```

chrisdp replied:
When trying to move to the async version I ran into another problem. I attempted to understand the issue but sadly I wasn't able to figure it out. This is following the import change in the comment above.
![image](https://user-images.githubusercontent.com/9591618/59079604-e774d900-88ba-11e9-8bec-4acb2083dea2.png)

TwitchBronBron commented on src/BrightScriptDebugSession.ts:
```diff
@@ -287,6 +306,157 @@ export class BrightScriptDebugSession extends DebugSession {
+                '.zip': 'application/zip'
+            };
+
+            http.createServer((req, res) => {
```
Can you break the http server stuff out into its own class?

TwitchBronBron commented on src/BrightScriptDebugSession.ts:
```diff
@@ -287,6 +306,157 @@ export class BrightScriptDebugSession extends DebugSession {
+                    });
+                });
+
+            }).listen(port);
```
What happens if we restart the debug session (instead of stop and start). Does the http server get completely discarded and completely recreated? I'm just concerned that a restart will error out because the http server and port are already occupied.

chrisdp replied:
Have not had any problems with the server and or port being occupied after debugger stops or restarts.

chrisdp commented:
In addressing feedback I discovered a problem with breakpoints that I will need to address. It seems like the breakpoints are no longer being added into the source channel. This is related to the logic in the `deleteBreakpointsForClientPath` function. I will look into this after addressing the more simple feedback.

chrisdp commented:
I managed to fix the breakpoint issue and other feedback. We are good for another round of review 😃

TwitchBronBron commented on package.json:
```diff
@@ -156,6 +157,30 @@
+                                "default": "${workspaceFolder}/libs"
+                            },
+                            "componentLibraries": {
+                                "type": "array",
```
Sorry for not catching this earlier. Using type "array" isn't good enough here, because the user won't get any intellisense for `rootDir`, `outFile`, `files`, etc. You should copy the `items` section for `files` down below into this. Then test and verify that in vscode, it provides intellisense for each of those items.

TwitchBronBron commented:
@chrisdp This looks great! Fantastic work. I just have the one comment, (adding json schema info for the componentLibraryFiles" section, and then we're good to go.

TwitchBronBron commented:
@chrisdp Oh, another thing. Can you please update the readme with your new debug config settings, as well as a section explaining how this stuff works? If the user doesn't know about it...then it doesn't exist. :D

chrisdp commented:
@TwitchBronBron I will update the README on usage first thing tomorrow. I have already pushed up the schema you requested so feel free to take a look at that.
